### DISCUSSION TITLE
Accessing the database from outside of main.go (package main)

### DISCUSSION THREAD

**Discussion by stamler** [pocketbase/pocketbase#5034] [Q&A] [answered]
I have a package called hooks across multiple files that I’m importing into my main package for code cleanliness (framework in Go)

Some of my hooks such as `OnRecordBeforeCreateRequest` require that I query the database for a row from a different collection prior to creating the new record. I understand the correct way to access the db is with `app.Dao().FindRecordById()`. How am I supposed to access this from within the functions in the `hooks` package? Is the recommended practice to simply pass the `app` instantiated with `app := pocketbase.New()` in `package main` into these functions in the hooks package as an argument? Is there another way to grab a reference to the existing Dao without having to pass it through the chain of calls?

**Comment by ganigeorgiev** [comment, marked as answer]
> Is the recommended practice to simply pass the app instantiated with app := pocketbase.New() in package main into these functions in the hooks package as an argument?

There are no hard recommendations. Like any other Go code it is up to you and your specific use case to decide how you want to organize it (_if you need a reference you could check https://github.com/presentator/presentator, but again this is just one example_).

In general, both options, passing the app instance or `Dao`, are fine depending on what actually your hooks functions do (eg. do they need access to the app settings, filesystem abstraction, mailer, etc.). Note that with the [refactoring](https://github.com/pocketbase/pocketbase/discussions/4355) the `core.App` and `Dao` will be merged. Additionally all app event hooks will now have `e.App` as field which would allow for example starting a transaction and executing the hook finalizer as part of it. So with the next release more-or-less passing the app as context arg will be encouraged.

> Is there another way to grab a reference to the existing Dao without having to pass it through the chain of calls?

You can call one of [Dao factories](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.22.13/daos#New) but you still need to pass a `dbx.Builder` instance to construct it so I don't see much benefit of doing that (_the db model event hooks will also not get triggered out of the box this way_).

**Reply by stamler** [reply]
Thanks for this helpful answer. PocketBase is excellent.
